test/UnitTestFramework/UnitTestFramework.js

```diff
@@ -167,8 +191,120 @@ var testRunner = function testRunner() {
+                        if (!passed) helpers.writeln(asyncTest.error[i]);
+                        helpers.writeln(passed ? "PASSED" : "FAILED");
+                    }
+                    !passed || passedTestCount++;
+                }
+                if (_verbose || executedTestCount != passedTestCount) {
+                    helpers.writeln("Summary of tests: total executed: ", executedTestCount,
+                      "; passed: ", passedTestCount, "; failed: ", executedTestCount - passedTestCount);
+                } else {
+                    helpers.writeln("pass");
+                }
+            });
+        },
+
+        runAsyncTest: function runAsyncTest(testIndex, testName, testBody) {
+            ///<summary>Runs test body catching all exceptions.<br/>
+            /// Result: prints LOADED/LOADING FAILED to the output.
+            ///</summary>
+            function logTestNameIf(b) {
+                if (b) {
+                    helpers.writeln("*** Loading test #", executedTestCount + 1, " (", testIndex, "): ", testName);
+                }
+            }
+
+            logTestNameIf(_verbose);
+
+            asyncTest.testIndex = testIndex;
+            asyncTest.promise[testIndex] = [];
+            asyncTest.resolve[testIndex] = [];
+
+            var isSuccess = true;
+            try {
+                testBody(testIndex);
+            } catch (ex) {
+                var message = ex.stack || ex.message || ex;
+                logTestNameIf(!_verbose);
+                helpers.writeln("Test threw exception: ", message);
+                isSuccess = false;
+            }
+            if (isSuccess) {
+                if (_verbose) {
+                    helpers.writeln("LOADED");
+                }
+                ++passedTestCount;
+            } else {
+                helpers.writeln("LOADING FAILED");
+                testRunner.asyncTestErr(testIndex, "LOADING FAILED");
+            }
+            ++executedTestCount;
+        },
+
+        asyncTestBegin : function asyncTestBegin(testIndex, testCount) {
+            _asyncMode = true;
+            asyncTest.testIndex = testIndex;
+        },
+
+        asyncTestEnd : function asyncTestEnd(testIndex, testCount) {
+            _asyncMode = false;
+            asyncTest.resolve[testIndex][testCount]();
+        },
+
+        async : function async(source, shouldFail) {
+            var testIndex = asyncTest.testIndex;
+            if (!shouldFail) {
+                var testCount = asyncTest.count[testIndex]++;
+                var promise = new Promise((resolve, reject) => {
+                    asyncTest.resolve[testIndex].push(resolve);
+                });
+                asyncTest.promise[testIndex].push(promise);
+                return 'testRunner.asyncTestBegin(' + testIndex + ", " + testCount + ');' + source + ";\ntestRunner.asyncTestEnd(" + testIndex + ", " + testCount + ");";

```

Comment:
String template would also be pretty nice here. 👍 

But actually, would it make more sense to have a helper here instead of requiring consumers of this API call `WScript.LoadModule(testRunner.async(...))` and munging up the module source code? Could be `testRunner.runModule(source, shouldFail)` which just loads/runs the module inline, returns a promise which does that, or returns a lambda which would do the same? Not sure I understand why we need to edit the module code itself here since that seems more complicated for new tests.